### ISSUE TITLE
Linq.Expressions.LambdaExpression and Linq.Expressions.BlockExpression

### ISSUE THREAD

**Issue by Warpten** [dotnet/corefx#26319]
Greetings;

Currently, `Linq.Expressions.BlockExpression` allows a set of local variables represented by `Linq.Expressions.ParameterExpression` to be provided to it.  This allows for the following "pseudo"-code:

```c#
var paramExpr = Expression.Parameter(typeof(object), "dummy");
var block = Expression.Block(new[] { paramExpr }, Expression.Call(paramExpr, typeof(object).GetMethod("GetHashCode", BindingFlags.Public | BindingFlags.Instance)));
var lambda = Expression.Lambda<Func<object, int>(block, paramExpr);
var fn = lambda.Compile();

Console.WriteLine("{0}", fn(new object()));
```

This code basically tries to call `object.GetHashCode()`. It however makes a critical mistake: `paramExpr` is declared as both a local variable to the block, and an argument to the compiled lambda. It will trigger an NPE.

`paramExpr` being declared as a variable within `block` is effectively shadowing the value of the argument, producing the same result as if it was assigned back to `null` - except that in this case, it's a local uninitialized variable, and Linq.Expressions is perfectly fine with that, as it should, since there's no optimization pass here.

The correct code would read:

```c#
var paramExpr = Expression.Parameter(typeof(object), "dummy");
var block = Expression.Block(Expression.Call(paramExpr, typeof(object).GetMethod("GetHashCode", BindingFlags.Public | BindingFlags.Instance)));
/// ... Or just get rid of the block entirely, obviously - I'm leaving it here for the sake of clarity
var lambda = Expression.Lambda<Func<object, int>(block, paramExpr);
var fn = lambda.Compile();

Console.WriteLine("{0}", fn(new object()));
```

Should `Expressions.Block` forbid usage of variable expressions that were created through `Expression.Parameter` as local variables? It seems to me that the difference between variables and parameters in expressions is just related to `Expression.Variable` and `Expression.Parameter`: semantically, I guess they are handled identically

**Comment by JonHanna** [comment]
> Should `Expressions.Block` forbid usage of variable expressions that were created through `Expression.Parameter` as local variables,

No. Because then you're banning use that works like:

```C#
var paramExpr = Expression.Parameter(typeof(object), "dummy");
var block = Expression.Block(new[] { paramExpr },
	Expression.Assign(paramExpr, Expression.Constant("test")),
	Expression.Call(paramExpr, typeof(object).GetMethod("GetHashCode", BindingFlags.Public | BindingFlags.Instance)));
var lambda = Expression.Lambda <Func<int>>(block);
var fn = lambda.Compile();
Console.WriteLine(fn());
```

> It seems to me that the difference between variables and parameters in expressions is just related to `Expression.Variable` and `Expression.Parameter`: semantically, I guess they are handled identically.

The only practical difference in implementation is that if you try to create a `ref` `Expression.Variable()` with `Expression.Variable(typeof(int).MakeByRefType())` it throws an exception. Aside from that, that `Expression.Variable()` in name better reflects an attempt to create a variable for use within a scope and `Expression.Parameter()` in name better reflects an attempt to create a parameter into a function is the only difference; a difference that's conceptual in the mind of the coder rather than reflected in the implementation.

People do though often use one for the other, so forbidding it now would be a breaking change.

One could make an argument for not allowing block variables that were also parameters to a containing lambda, but one can create valid code that does this and works, so it would still be a breaking change.

In all, Expressions is best considered as akin to a language that allows variable names to be redefined in inner scopes, the way C does, and your bug was to do so incorrectly.

Note that having `Variable` and `Parameter` doesn't completely remove the risk, anyway:

```C#
var paramExpr = Expression.Variable(typeof(object), "dummy");
var block = Expression.Block(new[] { paramExpr },
	Expression.Assign(paramExpr, Expression.Constant("test")),
	Expression.Block(new[] { paramExpr },
	Expression.Call(paramExpr, typeof(object).GetMethod("GetHashCode", BindingFlags.Public | BindingFlags.Instance))));
var lambda = Expression.Lambda <Func<int>>(block);
```

This uses `Variable` solely as variables for `Block` but still manages to have the same bug that you had, because it redefines it in an inner scope and doesn't assign it.

> it's a local uninitialized variable, and Linq.Expressions is perfectly fine with that, as it should, since there's no optimization pass here.

Strictly its not uninitialized, but unset, and has the default value for its type (akin to the way fields in C# get given default values without explicit initialization). There are no uninitialised values in expressions.

**Comment by Warpten** [comment]
Right, I forgot about the possible semantic misuse of Parameter as a Variable. I'm also coming from a world where breaking changes aren't exactly a big deal, but I can see why that would be the case for .NET.

I guess i just find it very odd to have two semantic usages of a same type, in what is basically an expression tree.

**Comment by JonHanna** [comment]
Life certainly is easier when one can move a bit faster and break more things 😄 

In a way, `ParameterExpression` isn't so much an expression that represents a parameter as an expression that represents a local value (and hence, a parameter or a variable). Perhaps it would have been called `LocalExpression` or something similar if `Block()` was in S.L.Expressions from its beginning with .NET 3.5 rather than added (along with `Variable()` which promises a result that can be used with it) later with .NET 4.0, but as it went there was once a time when the only sort of locals in expressions were indeed parameters.

**Comment by OmarTawfik** [comment]
@JonHanna @Warpten I'm reading this as no-op in that case in order to not break compatibility. Perhaps we can throw a better exception if the wrong type of `Expression` was used?

**Comment by JonHanna** [comment]
It's the same type of expression either way, and not even necessarily wrong to reuse a ParameterExpression, not assign to it (esp. if another branch would assign) and then use its value, so I don't think this could be usefully caught.

**Comment by Warpten** [comment]
I don't see a way to catch a use case that would not block the other, so I'm really not sure how to tackle this either. I had to take a step back from C# for a while so these notifications flew under the radar...

Maybe, when this `ParameterExpression` is first found in the tree, flag it as declared within a scope, and when detected as scope-local again (barring subcalls), throw? I think this is easier than just trying to walk back up the tree, which is probably not even supported, or would require an extra loop. (I might be missing use cases, I need to read up on this thread again)

**Comment by JonHanna** [comment]
That would throw for the working case I gave.

**Comment by Warpten** [comment]
Give me a minute, I need to read it again ;)
